rebasehelper/tags.py

```diff
@@ -80,7 +80,52 @@ def _map_sections_to_parsed(cls, raw_content: SpecContent, parsed_content: SpecC
 
     @classmethod
     def _parse(cls, raw_content: SpecContent, parsed_content: SpecContent) -> Tuple[Tag]:
-        """Parses all tags from provided SPEC content and determines if they are valid.
+        result = []
+        parsed_mapping = cls._map_sections_to_parsed(raw_content, parsed_content)
+        for section_index, (section, section_content) in enumerate(raw_content.sections):
+            section = section.lower()
+            parsed_index = parsed_mapping[section_index]
+            parsed: List[str] = [] if parsed_index == -1 else parsed_content.sections[parsed_index][1]
+            if section.startswith('%package'):
+                result.extend(cls._parse_package_tags(section, section_content, parsed))
+            elif section.startswith('%sourcelist') or section.startswith('%patchlist'):
+                result.extend(cls._parse_list_tags(section, section_content, parsed, cast(Tuple[Tag], tuple(result))))
+        return cast(Tuple[Tag], tuple(result))
+
+    @classmethod
+    def _parse_list_tags(cls, section: str, section_content: List[str], parsed: List[str],
+                         parsed_tags: Tuple[Tag]) -> List[Tag]:
+        """Parses all tags from a %sourcelist or %patchlist section.
+
+        Only parses tags that are valid (AKA are in parsed), nothing more can
+        consistently be detected.
+
+        Follows how rpm works, the new Source/Patch tags are indexed starting from
+        the last parsed Source/Patch tag.
+
+        """
+        tag = 'Source' if section == '%sourcelist' else 'Patch'
+        indexes = []
+        for parsed_tag in cls._filter(parsed_tags, name=tag + '*'):
+            try:
+                indexes.append(int(parsed_tag.name.lstrip(tag)))
+            except ValueError:
+                continue
+        index = 0 if not indexes else max(indexes) + 1
+        result = []
+        for i, line in enumerate(section_content):
+            expanded = MacroHelper.expand(line)
+            is_comment = SpecContent.get_comment_span(line, section)[0] != len(line)
+            if not expanded or not line or is_comment or not [p for p in parsed if p == expanded.rstrip()]:
+                continue
+            result.append(Tag(section, i, tag + str(index), (0, len(line) - 1), True))
+            index += 1
+
+        return result
+
+    @classmethod
+    def _parse_package_tags(cls, section: str, section_content: List[str], parsed: List[str]) -> List[Tag]:
+        """Parses all tags from provided by a %package section and determines if they are valid.
```

Comment:
I guess `from provided by` is a typo. I would rather use `in a %package section`.

Reply:
And the same above, to stay consistent.
BTW, why is `_parse_package_tags()` below `_parse_list_tags()`? It would feel more natural if their order was opposite.